common/network/queues/concurrent_queue.h.inc

```diff
@@ -0,0 +1,94 @@
+#include "common/network/cond_handle.h"
+#include "common/network/lock_handle.h"
+
+#define DECLARE_STRUCT_CONCURRENT_QUEUE_NODE_OF(TYPE)  \
+  typedef struct concurrent_queue_node_of_##TYPE##_s { \
+    TYPE data;                                         \
+    struct concurrent_queue_node_of_##TYPE##_s *next;  \
+  } concurrent_queue_node_of_##TYPE##_t;
+
+#define DECLARE_VTABLE_CONCURRENT_QUEUE_OF(TYPE)                           \
+  struct concurrent_queue_of_##TYPE##_s;                                   \
+  typedef struct vtable_concurrent_queue_of_##TYPE##_s {                   \
+    TYPE *(*front)(struct concurrent_queue_of_##TYPE##_s * this);          \
+    TYPE *(*back)(struct concurrent_queue_of_##TYPE##_s * this);           \
+    int (*empty)(struct concurrent_queue_of_##TYPE##_s * this);            \
+    size_t (*size)(struct concurrent_queue_of_##TYPE##_s * this);          \
+    void (*push)(struct concurrent_queue_of_##TYPE##_s * this, TYPE data); \
+    void (*pop)(struct concurrent_queue_of_##TYPE##_s * this);             \
+  } vtable_concurrent_queue_of_##TYPE##_t;                                 \
+  extern vtable_concurrent_queue_of_##TYPE##_t const                       \
+      vtable_concurrent_queue_of_##TYPE##_g;
+
+#define DECLARE_STRUCT_CONCURRENT_QUEUE_OF(TYPE)         \
+  typedef struct concurrent_queue_of_##TYPE##_s {        \
+    concurrent_queue_node_of_##TYPE##_t *front;          \
+    concurrent_queue_node_of_##TYPE##_t *back;           \
+    vtable_concurrent_queue_of_##TYPE##_t const *vtable; \
+    size_t size;                                         \
+    lock_handle_t lock;                                  \
+    cond_handle_t cond;                                  \
+  } concurrent_queue_of_##TYPE;
+
+/**
+ * Returns pointer to the first element in the queue. This element will be the
+ * first element to be removed on a call to pop()
+ *
+ * @return Pointer to the first element
+ */
+#define DECLARE_FUNCTION_FRONT_CONCURRENT_QUEUE_OF(TYPE) \
+  TYPE *CONCURRENT_QUEUE_OF_##TYPE##_FRONT(              \
+      struct concurrent_queue_of_##TYPE##_s *this);
```

Comment:
How about changing `this` to `thiz` in order to make C++ compilers happy?

Reply:
I'll call it self, thanks